src/utils/beacon/useHandleBeaconMessage.tsx

```diff
@@ -0,0 +1,151 @@
+import {
+  BeaconMessageType,
+  BeaconRequestOutputMessage,
+  PartialTezosOperation,
+  TezosOperationType,
+} from "@airgap/beacon-wallet";
+import { useContext } from "react";
+
+import { PermissionRequestModal } from "./PermissionRequestModal";
+import { SignPayloadRequestModal } from "./SignPayloadRequestModal";
+import { DynamicModalContext } from "../../components/DynamicModal";
+import { BatchSignPage } from "../../components/SendFlow/Beacon/BatchSignPage";
+import { BeaconSignPage } from "../../components/SendFlow/Beacon/BeaconSignPage";
+import { ImplicitAccount } from "../../types/Account";
+import { ImplicitOperations } from "../../types/AccountOperations";
+import { isValidContractPkh, parseContractPkh, parseImplicitPkh } from "../../types/Address";
+import { Operation } from "../../types/Operation";
+import { useGetOwnedAccount } from "../hooks/getAccountDataHooks";
+import { useSelectedNetwork } from "../hooks/networkHooks";
+import { useAsyncActionHandler } from "../hooks/useAsyncActionHandler";
+import { estimate } from "../tezos";
+
+/**
+ * @returns a function that handles a beacon message and opens a modal with the appropriate content
+ *
+ * For operation requests it will also try to convert the operation(s) to our {@link Operation} format,
+ * estimate the fee and open the BeaconSignPage only if it succeeds
+ */
+export const useHandleBeaconMessage = () => {
+  const { openWith } = useContext(DynamicModalContext);
+  const { handleAsyncAction } = useAsyncActionHandler();
+  const getAccount = useGetOwnedAccount();
+  const network = useSelectedNetwork();
+
+  return (message: BeaconRequestOutputMessage) => {
+    void handleAsyncAction(
+      async () => {
+        let modal;
+        switch (message.type) {
+          case BeaconMessageType.PermissionRequest: {
+            modal = <PermissionRequestModal request={message} />;
+            break;
+          }
+          case BeaconMessageType.SignPayloadRequest: {
+            modal = <SignPayloadRequestModal request={message} />;
+            break;
+          }
+          case BeaconMessageType.OperationRequest: {
+            const signer = getAccount(message.sourceAddress) as ImplicitAccount;
```

Comment:
There was an error modal in the past - are we just throwing an error now when signer is not an owned account?

Reply:
If so - can we have a test for this case please?